runhouse/rns/function.py

```diff
@@ -95,54 +88,58 @@ def _check_for_child_configs(cls, config):
 
     def to(
         self,
-        system: Union[str, Cluster],
+        system: Union[str, Cluster] = None,
+        env: Union[List[str], Env] = None,
+        # Variables below are deprecated
         reqs: Optional[List[str]] = None,
-        setup_cmds: Optional[List[str]] = None,
+        setup_cmds: Optional[List[str]] = [],
     ):
         """
-        Set up a Function on the given system, install the reqs, and run setup_cmds.
+        Set up a Function and Env on the given system.
 
         See the args of the factory method :func:`function` for more information.
         """
+        if setup_cmds:
+            warnings.warn(
+                "``setup_cmds`` argument has been deprecated. "
+                "Please pass in setup commands to the ``Env`` class corresponding to the function instead."
+            )
+
+        if reqs is not None:
+            warnings.warn(
+                "``reqs`` argument has been deprecated. Please use ``env`` instead."
+            )
+            env = Env(reqs=reqs, setup_cmds=setup_cmds)
+        elif env and isinstance(env, List):
+            env = Env(reqs=env, setup_cmds=setup_cmds)
+        else:
+            env = env or self.env
+
         # We need to backup the system here so the __getstate__ method of the cluster
         # doesn't wipe the client and _grpc_client of this function's cluster when
         # deepcopy copies it.
         hw_backup = self.system
         self.system = None
         new_function = copy.deepcopy(self)
         self.system = hw_backup
-        if isinstance(system, str):
-            system = Cluster.from_name(system, dryrun=self.dryrun)
-        new_function.system = system
-        new_function.setup_cmds = setup_cmds if setup_cmds else self.setup_cmds
 
-        reqs = reqs if reqs else self.reqs
-        new_reqs = []
-        for req in reqs:
-            # convert local packages to a package on the cluster
-            if isinstance(req, str) and Path(req.split(":")[-1]).expanduser().exists():
-                req = Package.from_string(req)
-                req = req.to(system)
-            new_reqs.append(req)
-        new_function.reqs = new_reqs
+        if system:
+            if isinstance(system, str):
+                system = Cluster.from_name(system, dryrun=self.dryrun)
+            new_function.system = system
+        else:
+            new_function.system = self.system
 
         logging.info("Setting up Function on cluster.")
-        new_function.system.install_packages(new_function.reqs)
-        if self.setup_cmds:
-            new_function.system.run(self.setup_cmds)
+        new_env = env.to(new_function.system)
         logging.info("Function setup complete.")
+        new_function.env = new_env
 
         return new_function
 
-    def run_setup(self, cmds: List[str], force: bool = False):
+    def run_setup(self, cmds: List[str]):
```

Comment:
I think was just thinking about it as a command to run once a cluster, not necessarily corresponding to a function vs env. In the case you mentioned, I think it would make more sense to leave it as part of the function, but to run the command in the env on the cluster when we send it to env, cluster?

Reply:
Ah ok. Ya, I think the logic below is a little galaxy brain. Let's leave it how you have it, and when we add runs we should have an obvious way to say "run this once on the cluster and not over and over each time I run my script.